### PR TITLE
Add new feature to fix invalid constraints.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to recent compiler work to detect this mistake and report a custom diagnostic for it.

Looks like this:

image

### CODE REVIEW

**Comment by jcouv** [comment]
I'm surprised this is targeting a specific release branch. Why not target `main` branch as usual?

File: .github/CODEOWNERS
Comment (jcouv) [inline .github/CODEOWNERS:11]: This should not be in this PR. It will be removed from `main` anyways. Likely related to PR targeting `release/dev17.3` instead of `main`.

File: src/Analyzers/CSharp/CodeFixes/FixIncorrectConstraint/CSharpFixIncorrectConstraintCodeFixProvider.cs
Comment (jcouv) [inline src/Analyzers/CSharp/CodeFixes/FixIncorrectConstraint/CSharpFixIncorrectConstraintCodeFixProvider.cs:48]: Just to confirm, most of the cases returning `false` seem unreachable. This is just being defensive, right?

File: src/Compilers/Core/Portable/Compilation/Compilation.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:1174]: Ditto, this change doesn't belong here

File: src/EditorFeatures/CSharpTest/FixIncorrectConstraint/FixIncorrectConstraintTests.cs
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/FixIncorrectConstraint/FixIncorrectConstraintTests.cs:54]: nit: Blank line above `RunAsync`. Also in other tests

File: src/EditorFeatures/CSharpTest/RawStringLiteral/RawStringLiteralCommandHandlerTests.cs
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/RawStringLiteral/RawStringLiteralCommandHandlerTests.cs:122]: Not sure whether this is meant to be in this PR, or an artifact of targeting 17.3 branch.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs
Comment (jcouv) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs:373]: Same here

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 10)

**Comment by CyrusNajmabadi** [comment]
> I'm surprised this is targeting a specific release branch. Why not target main branch as usual?

@jcouv I agree. But new features are not allowed in main currently. They have to go into 17.3

File: src/Analyzers/CSharp/CodeFixes/FixIncorrectConstraint/CSharpFixIncorrectConstraintCodeFixProvider.cs
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/FixIncorrectConstraint/CSharpFixIncorrectConstraintCodeFixProvider.cs:48]: Correct.  Compiler diagnostics come with no contracts/guarantees. They can and do change over time. So consumption code needs to be written to be resilient to that.

**Comment by CyrusNajmabadi** [comment]
The other stuff got pulled in with retargetting. Once main is merged in, they'll go away

**Review by jcouv — COMMENTED** [review]
LGTM, but I'll wait until the extra changes no longer appear in diff to sign-off. Thanks

File: src/EditorFeatures/CSharpTest/FixIncorrectConstraint/FixIncorrectConstraintTests.cs
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/FixIncorrectConstraint/FixIncorrectConstraintTests.cs:25]: sure.  happy to add.

**Review by davidwengier — APPROVED** [review]

**Pull request merged**
